File: JavaScript/test/src/bucket_replication.js
// 设置存储桶跨地域复制规则
function putBucketReplication(assert) {
  //.cssg-snippet-body-start:[put-bucket-replication]
  cos.putBucketReplication({
      Bucket: 'examplebucket-1250000000',  /* 必须 */
      Region: 'COS_REGION',     /* 存储桶所在地域，必须字段 */
      ReplicationConfiguration: { /* 必须 */
          Role: "qcs::cam::uin/100000000001:uin/100000000001",
          Rules: [{
              ID: "1",
              Status: "Enabled",
              Prefix: "sync/",
              Destination: {
                  Bucket: "qcs::cos:ap-beijing::destinationbucket-1250000000",
                  StorageClass: "Standard",
              }
          }]
      }
  }, function (err, data) {
      console.log(err || data);
  });
  
  //.cssg-snippet-body-end
}

// 获取存储桶跨地域复制规则
function getBucketReplication(assert) {
  //.cssg-snippet-body-start:[get-bucket-replication]
  cos.getBucketReplication({
      Bucket: 'examplebucket-1250000000', /* 必须 */
      Region: 'COS_REGION',     /* 存储桶所在地域，必须字段 */
  }, function(err, data) {
      console.log(err || data);
  });
  
  //.cssg-snippet-body-end
}

// 删除存储桶跨地域复制规则
function deleteBucketReplication(assert) {
  //.cssg-snippet-body-start:[delete-bucket-replication]
  cos.deleteBucketReplication({
      Bucket: 'examplebucket-1250000000', /* 必须 */
      Region: 'COS_REGION',     /* 存储桶所在地域，必须字段 */
  }, function(err, data) {
      console.log(err || data);
  });
  
  //.cssg-snippet-body-end
}

//.cssg-methods-pragma

test("BucketReplication", async function(assert) {
  // 设置存储桶跨地域复制规则
  await putBucketReplication(assert)

  // 获取存储桶跨地域复制规则
  await getBucketReplication(assert)

  // 删除存储桶跨地域复制规则
  await deleteBucketReplication(assert)

//.cssg-methods-pragma
})

File: Go/examples/bucket_replication.go
package cos

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"strings"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/suite"
	"github.com/tencentyun/cos-go-sdk-v5"
)

type CosTestSuite struct {
	suite.Suite
	VariableThatShouldStartAtFive int

	// CI client
	Client *cos.Client

	// Copy source client
	CClient *cos.Client

	// test_object
	TestObject string

	// special_file_name
	SepFileName string
}

var (
	UploadID string
	PartETag string
)

// 设置存储桶跨地域复制规则
func (s *CosTestSuite) putBucketReplication() {
  client := s.Client
  //.cssg-snippet-body-start:[put-bucket-replication]
  opt := &cos.PutBucketReplicationOptions{
      // qcs::cam::uin/[UIN]:uin/[Subaccount]
      Role: "qcs::cam::uin/100000760461:uin/100000760461",
      Rule: []cos.BucketReplicationRule{
          {
              ID: "1",
              // Enabled or Disabled
              Status: "Enabled",
              Destination: &cos.ReplicationDestination{
                  // qcs::cos:[Region]::[Bucketname-Appid]
                  Bucket: "qcs::cos:ap-beijing::destinationbucket-1250000000",
              },
          },
      },
  }
  _, err := client.Bucket.PutBucketReplication(context.Background(), opt)
  if err != nil {
      panic(err)
  }
  
  //.cssg-snippet-body-end
}

// 获取存储桶跨地域复制规则
func (s *CosTestSuite) getBucketReplication() {
  client := s.Client
  //.cssg-snippet-body-start:[get-bucket-replication]
  _, _, err := client.Bucket.GetBucketReplication(context.Background())
  if err != nil {
      panic(err)
  }
  
  //.cssg-snippet-body-end
}

// 删除存储桶跨地域复制规则
func (s *CosTestSuite) deleteBucketReplication() {
  client := s.Client
  //.cssg-snippet-body-start:[delete-bucket-replication]
  _, err := client.Bucket.DeleteBucketReplication(context.Background())
  if err != nil {
      panic(err)
  }
  
  //.cssg-snippet-body-end
}

//.cssg-methods-pragma


func TestCOSTestSuite(t *testing.T) {
	suite.Run(t, new(CosTestSuite))
}

func (s *CosTestSuite) TestBucketReplication() {
	// 将 examplebucket-1250000000 和 ap-guangzhou 修改为真实的信息
	u, _ := url.Parse("https://examplebucket-1250000000.cos.ap-guangzhou.myqcloud.com")
	b := &cos.BaseURL{BucketURL: u}
	c := cos.NewClient(b, &http.Client{
		Transport: &cos.AuthorizationTransport{
			SecretID:  os.Getenv("COS_KEY"),
			SecretKey: os.Getenv("COS_SECRET"),
		},
	})
  s.Client = c

	// 设置存储桶跨地域复制规则
	s.putBucketReplication()

	// 获取存储桶跨地域复制规则
	s.getBucketReplication()

	// 删除存储桶跨地域复制规则
	s.deleteBucketReplication()

	//.cssg-methods-pragma
}
